### PR TITLE
ref(poc): composition

### PR DESCRIPTION
**Pull request by JonasBa**
This is a proof of concept PR that is not meant to be merged, I created it as a showcase on how leveraging composition can help reduce API surface and increase flexibility of the codebase. 

The changes in the PR are the following
- removal of showPermalink boolean prop
- creating a PermalinkTitle component
- passing a PermalinkTitle or Title component based on what used to be showPermalink

With the changes introduced here, you can see that we have removed one prop from the component while simultaneously also exposed the api of PermalinkTitle which was previously inaccessible to the consumer, meaning we can now independently pass custom props to each instance of the PermalinkTitle component without having to extend the API of our TraceEvent component and avoid prop drilling entirely. 

To summarize the functional benefits:
- You are no longer limited to Title or PermalinkTitle - you can pass whatever component you like
- Exposes PermalinkTitle meaning you can pass different props to each instance however you please to without extending/adding new props
- Reduced API surface (1 prop removed)

In an ideal scenario, the PermalinkTitle would have been a core UI library component so as to avoid creating new custom components around the codebase and the Title component would not have included any margins by default so that we would not have to override them.

Some resources:
- https://medium.com/dailyjs/applying-solid-principles-in-react-14905d9c5377 shared by @nhsiehgit 
- https://felixgerschau.com/react-component-composition/

### CODE REVIEW

**Review by markstory — COMMENTED** [review]
Any reason you don't want to merge this? It looks like a positive change to me.

**Comment by JonasBa** [comment]
@markstory I havent really verified that it works _exactly_ as before, was mostly meant to demonstrate what I meant

File: static/app/components/events/traceEventDataSection.tsx [hunk 1]
```diff
@@ -287,7 +289,8 @@ const StyledIconAnchor = styled(IconAnchor)`
 const Permalink = styled('a')`
   display: inline-flex;
   justify-content: flex-start;
-  :hover ${StyledIconAnchor} {
+
+  &:hover ${StyledIconAnchor} {
```
Comment (priscilawebdev) [inline static/app/components/events/traceEventDataSection.tsx:293]: please, what is the difference between `&:hover ` and ` :hover`?

**Comment by priscilawebdev** [comment]
I like the idea too, but before we can merge it, we have to make sure that everything is according to the previous styles. 


Great PoC @JonasBa! Thanks for sharing this 👏

**Review by k-fish — APPROVED** [review]
Nice 👍, I agree about shipping it as an example (if it's not too much work to fix styles)

File: static/app/components/events/traceEventDataSection.tsx (hunk 1 above)
Comment (JonasBa) [reply static/app/components/events/traceEventDataSection.tsx:293]: @k-fish I actually replaced it thinking that it is broken and that it probably doesn't compile correctly 😅

File: static/app/components/events/traceEventDataSection.tsx [hunk 2]
```diff
@@ -277,6 +264,21 @@ export function TraceEventDataSection({
...
+  > {}
+
+export function PermalinkTitle(props: PermalinkTitleProps) {
+  return (
+    <Permalink href={'#' + props.type} className="permalink" {...props}>
+      <StyledIconAnchor />
+      {t('Stack Trace')}
+    </Permalink>
+  );
+}
```
[the remark this answers is not in the document]
Comment (JonasBa) [reply static/app/components/events/traceEventDataSection.tsx:280]: Ah yes 💯  you explained this perfectly, I was not really sure how to call it. I might have read that and just forgot about it

File: static/app/components/events/traceEventDataSection.tsx (hunk 1 above)
Comment (priscilawebdev) [reply static/app/components/events/traceEventDataSection.tsx:293]: oh, I see... can we enforce this through eslint, so we can keep the consistency? and thanks for the explanation @k-fish 🙏

**Comment by JonasBa** [comment]
@priscilawebdev let me know if you would have 10min to spare to pair on this PR and wrap it up. It's a bit hard for me to assert everything works as intended.

**Comment by priscilawebdev** [comment]
@JonasBa is this still a draft PR?

**Comment by JonasBa** [comment]
@priscilawebdev nope, I hadn't noticed I left it as a draft. It should be good to go, I'd appreciate if you can give it another look to make sure I haven't missed anything

File: static/app/components/events/traceEventDataSection.tsx [hunk 3]
```diff
@@ -54,9 +60,8 @@ type Props = {
   recentFirst: boolean;
   stackTraceNotFound: boolean;
   stackType: STACK_TYPE;
-  title: React.ReactNode;
+  title: React.ReactElement<any, any>;
```
Comment (priscilawebdev) [inline static/app/components/events/traceEventDataSection.tsx:63]: did we change this because it can be `null` or `undefined`?

**Review by priscilawebdev — APPROVED** [review]
checked locally and it looks good to me

File: static/app/components/events/traceEventDataSection.tsx (hunk 3 above)
Comment (JonasBa) [reply static/app/components/events/traceEventDataSection.tsx:63]: I changed it because you cant cloneElement on null/undefined and it didnt look like null/undefined is intended usage.
